Add constraints docs (#2943)
tarantool/doc: doc/book/box/data_model.rst

    |                                |                                           |                                      |
    +--------------------------------+-------------------------------------------+--------------------------------------+
    | ``'integer'``                  | :ref:`integer <index-box_integer>`,       | TREE or HASH                         |

>     | (may also be called ``'int'``) | which may include unsigned values         |                                      |

Reviewer: This is a small correction, but it already breaks the translation. Please introduce the same correction in the msgid in the corresponding `.po` file (`/locale/ru/LC_MESSAGES/book/box/data_model.po`).
More on this:
https://www.notion.so/tarantool/0165f40dd7ab4bb08889c2698ea25c34

in the
`Common Language Data Repository <https://unicode.org/cldr/charts/30/collation>`_.

.. _index-constraints:

-----------
Constraints
-----------

For better control over stored data, Tarantool supports **constraints** – user-defined
limitations on the values of certain fields or entire tuples. Together with data types,

> constraints allow to limit ranges of available field values both syntactically and semantically.

Reviewer: ```suggestion
constraints allow limiting the ranges of available field values both syntactically and semantically.
```
"Allow" requires a gerund. Another option is "allows _the user_ to ...", that is, there is an agent between "allow" and "to". Other options are considered ungrammatical: https://english.stackexchange.com/questions/60271/grammatical-complements-for-allow/60285#60285

For example, the field ``age`` typically has the ``number`` type, so it cannot store
strings or boolean values. However, it can still have values that don't make sense,
such as negative numbers. This is where constraints come to help.

.. _index-constraint_types:

****************
Constraint types
****************

There are two types of constraints in Tarantool:

* *field constraints* check that the value being assigned to a field
  satisfies a given condition. For example, ``age`` must be non-negative.

> * *tuple constraints* check complex conditions that can involve all fields of

Reviewer: ```suggestion
* *Field constraints* check that the value being assigned to a field
  satisfies a given condition. For example, ``age`` must be non-negative.

* *Tuple constraints* check complex conditions that can involve all fields of
```
Check the [list punctuation guideline](https://www.tarantool.io/en/doc/latest/contributing/docs/style/#lists).

  a tuple. For example, a tuple contains a date in three fields:
  ``year``, ``month``, and ``day``. You can validate ``day`` values based on
  the ``month`` value (and even ``year`` if you consider leap years).

> Field constraints work faster while tuple constraints allow implementing

Reviewer: ```suggestion
Field constraints work faster, while tuple constraints allow implementing
```

a wider range of limitations.

.. _index-constraint_functions:

********************
Constraint functions
********************

Constraints use stored Lua functions. They must return ``true`` when the constraint
is satisfied. Other return values (including ``nil``)and exceptions make the

> check fail and prevent tuple insertion or modification.

Reviewer: ```suggestion
Constraints use stored Lua functions, which must return ``true`` when the constraint
is satisfied. Other return values (including ``nil``) and exceptions make the
check fail and prevent tuple insertion or modification.
```
It is not immediately clear what "they" are.

> To create a constraint function, use :ref:`func.create with function body <box_schema-func_create_with-body>`:

Reviewer: ```suggestion
To create a constraint function, use :ref:`func.create with function body <box_schema-func_create_with-body>`.
```
Nested colons look awkward.
